dilation-protocol.md

```diff
@@ -1,371 +1,280 @@
-it as lost when bidirectional traffic cannot be maintained. It uses PING and
-PONG messages to detect this. These also serve to keep NAT entries alive,
-since many firewalls will stop forwarding packets if they don't observe any
-traffic for e.g. 5 minutes.
+and must be processed in a specific order
+(the Leader must not accept the Follower's message until it has generated its own).
+Noise allows handshake messages to include a payload, but we do not use this feature.
+
+All subsequent messages as known as "Noise transport messages", and are independent for each direction.
+Transport messages are encrypted by the shared key, in a form that evolves as more messages are sent.
+
+The Follower's first transport message is an empty packet, which we use as a "key confirmation message" (KCM).
+
+The Leader doesn't send a transport message right away: it waits to see the Follower's KCM,
+which indicates this connection is viable (i.e. the Follower used the same dilation key as the Leader,
+which means they both used the same wormhole code).
+
+Multiple connection attempts and handshakes may happen in parallel.
+Of those where the Leader received a KCM, it chooses one of them to be actually used.
+It will send an empty KCM to the Follower to mark that connection, and close all others.
+All other connection attempts on both sides will be cancelled.
+All listening sockets may or may not be shut down (TODO: think about it ~warner).
+
+### Messages over the Dilation connection
+
+Unlike with applications using Transit,
+the application layer messages are not sent directly over the Dilation connection.
+Instead, they get wrapped by this middle layer which provides the following features:
+
+- Sub-channel multiplexing
+- Pings and ACKs for connection status tracking
+- Automatic reconnection after connection loss
+- Durable channel with automatic re-transmission of lost messages after reconnect
+
+~~In the future, we might have L2 links that are less connection-oriented,
+which might have a unidirectional failure mode, at which point we'll need to
+monitor full roundtrips. To accomplish this, the Leader will send periodic
+unconditional PINGs, and the Follower will respond with PONGs. If the
+Leader->Follower connection is down, the PINGs won't arrive and no PONGs will
+be produced. If the Follower->Leader direction has failed, the PONGs won't
+arrive. The delivery of both will be delayed by actual data, so the timeouts
+should be adjusted if we see regular data arriving.~~
+
+If the Leader loses the Dilation connection for whatever reason, it will initiate a reconnection attempt.
+This involves sending a `reconnect` message and according hints over the Mailbox connection,
+the same way as for initially establishing the connection.
+Once a new connection is established and the handshake completes,
+both sides re-transmit their outbound messages that were lost due to the connection interruption.
```

Comment:
Careful here: these are L4 messages. If you instead re-transmitted L2 messages, the crypto may break (and another reason to use distinct keys for different generations).

Reply:
(I see you resolved this, but I don't seen changes .. maybe your workflow is different from mine).

This should also say how a peer determines which messages were lost (or potentially lost).

Reply:
Oops I'm sorry.

I usually keep track of open discussions as my TODO list (at least for things that are easily fixable), but I forgot to push my changes after doing so.